Break to classes

AA-Turner opened the pull request:
Very high level this breaks ScoutMap into 4 classes: ScoutData, Boundary, Map, and DistrictBoundaries

**Will write a full explanation on Sunday.** Hopefully this should both improve readability & maintainabilty, and make it easier to add more features & apply OOP.

This should also make it easier to start doing unit tests - I've got a local branch working on it.

Note: this was about 5 days work, and I broke the completed product into smaller commits that hopefully make sense thematically, though this does mean that inbetween commits the application will not work (it might do, but not guaranteed!)

Resolves #23 (it was easier putting it in here as I knew I was nearly finished)

Next steps are to split out HistorySummary, though I've only done the first function.

Thoughts comments queries? Appreciate that this is a massive change, so open to pushback! However I do think that it will make it easier to work on the application as a whole - a stupid one but my IDE can now immediatley tell me that something's wrong with type hinting etc when I'm writing a script

ox-baloo commented:
I've added a small commit, after trying to run a couple of the scripts, hope that's okay?

I really love this, thank you so much, yes it will make it easier, also makes the scripts more versitile (i.e. can have multiple ScoutData objects).

Couple of quick questions, would it be possible to think about the functions in utility.py, I just wondered whether a couple of them would be more suited to going into the classes more properly (is there an advantage to filter_records being in here).

Secondly, is to have a look at the commenting, in particular around base, I think some additional comments explaining what the functions are and how to use them might be helpful,

Finally, logging, it would be great to have a broadly consistent approach to logging. I heard one mantra that was every time there's a logical statement (i.e. if) there should be debug log, so that from the debug log you can track where the code goes. Another idea is to record the values of each function call in a human readable manner (filter_records) has a good example of this.

Thank you so much for doing this, I've just had another request, this time from the community impact team for data and maps, so I think word is getting around :)

AA-Turner commented:
Addressed in replies!

> I've added a small commit, after trying to run a couple of the scripts, hope that's okay?
Yes - thanks, bit embarresed I missed that! I accidentally reverted your change to import log_util methods direct in base.py as I was considering merging the two into one file, but then didn't. I don't mind having namespaces on function calls but don't know your thoughts?

> I really love this, thank you so much, yes it will make it easier, also makes the scripts more versitile (i.e. can have multiple ScoutData objects).
😊 

> Couple of quick questions, would it be possible to think about the functions in utility.py, I just wondered whether a couple of them would be more suited to going into the classes more properly (is there an advantage to filter_records being in here).
So the reason I put quite a lot of functions in here is that I think they can be class agnostic - at some point we may have another data class (CompassData?!), and I think a lot of the functions could apply to any dataset - it just happens that currently we only have one.
Something like section_from_type is only used in Map, so I could see that happening.
Another option is to add proxies to the utility functions in those classes? so ScoutCensus.years_of_return passes through to utility....

> Secondly, is to have a look at the commenting, in particular around base, I think some additional comments explaining what the functions are and how to use them might be helpful,
See additional commit f7ff468 - is this too little/enough/too much? Happy to add more!

> Finally, logging, it would be great to have a broadly consistent approach to logging. I heard one mantra that was every time there's a logical statement (i.e. if) there should be debug log, so that from the debug log you can track where the code goes. Another idea is to record the values of each function call in a human readable manner (filter_records) has a good example of this.
Not done this (yet)! A consistent approach to logging would be good, however there is I think the risk of over-logging and missing the code for the logs - I would say the debug process in your IDE would be better for debugging? Definitley open to adding more human-friendly logs!

> Thank you so much for doing this, I've just had another request, this time from the community impact team for data and maps, so I think word is getting around :)
Good good good!

ox-baloo commented:
Okideokie - I like the level of commenting looks good, I know basically understand. More importantly know how to maintain, and provided extra resources in case I break it!

And I like your forward view into class agnositism, good idea :)
